code/datums/components/memetic.dm

```diff
@@ -0,0 +1,126 @@
+//handles anything that has memetic properties and needs to keep track of affected humans. Memetics are designed to work only on humans.
+/datum/component/memetic
+	///List of affected humans
+	var/list/weakref/affected_mobs_weakref = list()
+	///Memetic bitflags to determine what should be counted as affected
+	var/memetic_flags
+	///Proc to run on affected humans
+	var/affected_proc
+	///List of sounds that count as being memetic
+	var/list/memetic_sounds
+
+/datum/component/memetic/Initialize(flags, meme_proc, memeticSounds)
+	. = ..()
+	memetic_flags = flags
+	affected_proc = meme_proc
+	memetic_sounds = memeticSounds
+
+/datum/component/memetic/Destroy(force, silent)
+	. = ..()
+	LAZYCLEARLIST(affected_mobs_weakref)
+
+/datum/component/memetic/RegisterWithParent()
+	RegisterSignal(parent, list(COMSIG_OBJECT_SOUND_HEARD, COMSIG_OBJECT_SOUND_HEARD_LOOPING), .proc/heard_memetic)
+	RegisterSignal(parent, COMSIG_ATOM_EXAMINED, .proc/examined_memetic)
+	RegisterSignal(parent, COMSIG_PHOTO_SHOWN_OF, .proc/saw_memetic_photo)
+	RegisterSignal(parent, COMSIG_ATOM_VIEW_RESET, .proc/saw_through_camera)
+
+/datum/component/memetic/UnregisterFromParent()
+	UnregisterSignal(parent, list(
+		COMSIG_OBJECT_SOUND_HEARD,
+		COMSIG_ATOM_EXAMINED,
+		COMSIG_PHOTO_SHOWN_OF,
+		COMSIG_ATOM_VIEW_RESET,
+		COMSIG_OBJECT_SOUND_HEARD_LOOPING
+	))
+
+/datum/component/memetic/proc/check_viewers() //I dont like doing this but since theres no way for us to send a signal upon something being viewed its neccesary
+	var/list/mviewers = viewers(world.view, parent)
+
+	for(var/mob/living/carbon/human/H in mviewers)
+		saw_memetic(H)
+
+/datum/component/memetic/proc/activate_memetic_effects()
+	if(!(memetic_flags & MPERSISTENT)) //if we arent a persistent memetic, then affected humans will be removed from effect after they no longer meet the reqs
+
+		for(var/weakref/Hweakref in affected_mobs_weakref)
```

Comment:
also you have the resolve weakrefs list proc, why are you doing it manually here?

Reply:
Because I am a foolish man bound by the constraints of my material form

(I made the resolve proc after writing this and forgor, will fix)